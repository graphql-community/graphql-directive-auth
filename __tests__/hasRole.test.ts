import { graphql } from 'graphql';
import MockExpressRequest from 'mock-express-request';
import AuthDirective from '../src/index';
import schema from '../example/schema';

beforeAll(() => {
  process.env.APP_SECRET = '123';
});

test('getDirectiveDeclaration should be defined', () => {
  expect(AuthDirective().hasRole.getDirectiveDeclaration()).toMatchSnapshot();
});

test('if throw error if no role inside token payload', () =>
  graphql(
    schema,
    `
      query {
        you
      }
    `,
    {},
    {
      req: new MockExpressRequest({
        headers: {
          Authorization:
            'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbG8iOjEyMywiaWF0IjoxNTMyNzI3MDMzfQ.nSGh5q9YXZd-pWYagOxD2f-9hdNPw08e4MnKASgn4wY',
        },
      }),
    }
  ).then(response => {
    expect(response.errors[0].message).toEqual(
      'Invalid token payload, missing role property inside!'
    );
    expect(response.data.you).toBeNull();
    expect(response).toMatchSnapshot({
      data: {
        you: null,
      },
      errors: expect.anything(),
    });
  }));

test('if role are correct', () =>
  graphql(
    schema,
    `
      query {
        you
      }
    `,
    {},
    {
      req: new MockExpressRequest({
        headers: {
          Authorization:
            'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InVzZXJfaWQiLCJyb2xlIjoiVVNFUiIsImlhdCI6MTUzMzA0ODk5N30.DISf4XuHkVo7YPXjY2VQWXgZge-c_ejLzsiBql2mXIs',
        },
      }),
    }
  ).then(response => {
    expect(response.errors).toBeUndefined();
    expect(response).toMatchSnapshot({
      data: {
        you: expect.any(String),
      },
    });
  }));

test('if throw an Error when user does not have correct roles', () =>
  graphql(
    schema,
    `
      query {
        together
      }
    `,
    {},
    {
      req: new MockExpressRequest({
        headers: {
          Authorization:
            'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InVzZXJfaWQiLCJyb2xlIjoiVVNFUiIsImlhdCI6MTUzMzA0ODk5N30.DISf4XuHkVo7YPXjY2VQWXgZge-c_ejLzsiBql2mXIs',
        },
      }),
    }
  ).then(response => {
    expect(response.errors[0].message).toEqual(
      'Must have role: MALINA, you have role: USER'
    );
    expect(response.data.together).toBeNull();
    expect(response).toMatchSnapshot({
      data: {
        together: null,
      },
      errors: expect.anything(),
    });
  }));

test('if return null for a field when does not have permissions', () =>
  graphql(
    schema,
    `
      query {
        field {
          public
          private
        }
      }
    `,
    {},
    {
      req: new MockExpressRequest({
        headers: {
          Authorization:
            'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InVzZXJfaWQiLCJyb2xlIjoiVVNFUiIsImlhdCI6MTUzMzA0ODk5N30.DISf4XuHkVo7YPXjY2VQWXgZge-c_ejLzsiBql2mXIs',
        },
      }),
    }
  ).then(response => {
    expect(response).toMatchSnapshot({
      data: {
        field: {
          public: 'public_exists',
          private: null,
        },
      },
    });
  }));
