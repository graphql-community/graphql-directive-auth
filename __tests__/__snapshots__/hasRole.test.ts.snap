// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDirectiveDeclaration should be defined 1`] = `
GraphQLDirective {
  "args": Array [
    Object {
      "astNode": undefined,
      "defaultValue": undefined,
      "description": null,
      "name": "role",
      "type": "String",
    },
  ],
  "astNode": undefined,
  "description": undefined,
  "locations": Array [
    "FIELD_DEFINITION",
  ],
  "name": "hasRole",
}
`;

exports[`if return null for a field when does not have permissions 1`] = `
Object {
  "data": Object {
    "field": Object {
      "private": null,
      "public": "public_exists",
    },
  },
}
`;

exports[`if role are correct 1`] = `
Object {
  "data": Object {
    "you": Any<String>,
  },
}
`;

exports[`if throw an Error when user does not have correct roles 1`] = `
Object {
  "data": Object {
    "together": null,
  },
  "errors": Anything,
}
`;

exports[`if throw error if no role inside token payload 1`] = `
Object {
  "data": Object {
    "you": null,
  },
  "errors": Anything,
}
`;
